### PR TITLE
.s2ifile restore multiple environment variables not working

### LINKED ISSUE #761: Setting multiple environment variables in .s2ifile does not work
Executing the following command work as expected:
s2i build . ubuntu:latest --scripts-url file://.s2i/exec -e "TEST_VAR1=testvar1" -e "TEST_VAR2=testvar2"

If I try to create an .s2ifile with multiple env vars by executing the same command with --use-config added it generates an .s2ifile with "flags":{"env":"TEST_VAR1=testvar1,TEST_VAR2=testvar2" ......}

After this executing "s2i build --use-config" creates a single TEST_VAR1 environment variable with value "testvar1,TEST_VAR2=testvar2" instead of two environment variables.

Also a warning message appears:
WARNING: DEPRECATED: Use multiple -e flags to specify multiple environment variables instead of comma ("TEST_VAR1=testvar1,TEST_VAR2=testvar2")

Environment attached.
Workaround (kindda): using environment file instead

[reproduce.zip](https://github.com/openshift/source-to-image/files/1102327/reproduce.zip)

### PR DESCRIPTION
**Pull request by coreydaley**
Restoring multiple environment variables from an .s2ifile
does not work correctly since removing the ability to supply
a comma separated list of environment variables on the cli.

This pull request restores the environment variables back into
a list before running the command when building the command from
a saved .s2ifile

Fixes #761

### CODE REVIEW

**Comment by coreydaley** [comment]
[test]

**Comment by coreydaley** [comment]
@openshift/devex ptal

**Comment by bparees** [comment]
@coreydaley we shouldn't be writing it out to the file in comma format either, because we lose fidelity if one of their env vars contained a comma.

**Comment by coreydaley** [comment]
@bparees ptal

**Comment by coreydaley** [comment]
@bparees ptal

File: pkg/config/config.go [hunk 1]
```diff
@@ -63,14 +74,31 @@ func Restore(config *api.Config, cmd *cobra.Command) {
...
 	config.BuilderImage = c.BuilderImage
 	config.Source = c.Source
 	config.Tag = c.Tag
+	cmd.Flags().Visit(func(f *pflag.Flag) {
+		if f.Name == "env" {
+			envOverride = true
+		}
+	})
 	for name, value := range c.Flags {
-		// Do not change flags that user sets. Allow overriding of stored flags.
```
Comment (jim-minter) [inline pkg/config/config.go]: suggest preserving the comment

File: pkg/config/config.go [hunk 2]
```diff
@@ -63,14 +74,31 @@ func Restore(config *api.Config, cmd *cobra.Command) {
...
-		if cmd.Flag(name).Changed {
-			continue
+		if name == "env" {
+			for _, v := range strings.Split(value, ",") {
```
Comment (jim-minter) [inline pkg/config/config.go]: cmd.Flag(name).Changed test needed here?

**Comment by jim-minter** [comment]
That moment when you realise your serialisation format is utterly inadequate :cry:
I think this is workable as a quick fix, but an unambiguously new Config struct (version field?) with something like a `Flags []struct { key, value string }` + migration for old versions might be better?

File: pkg/config/config.go [hunk 3]
```diff
@@ -26,6 +29,7 @@ type Config struct {
 
 // Save persists the S2I command line arguments into disk.
 func Save(config *api.Config, cmd *cobra.Command) {
+	glog.V(0).Infof("Saving")
```
Comment (bparees) [inline pkg/config/config.go]: debug?

File: pkg/config/config.go [hunk 4]
```diff
@@ -34,7 +38,13 @@ func Save(config *api.Config, cmd *cobra.Command) {
 	}
 	// Store only flags that have changed
```
Comment (bparees) [inline pkg/config/config.go]: this comment doesn't make sense.  isn't this saving all the flags?  (which seems fine)

File: pkg/config/config.go [hunk 5]
```diff
@@ -49,6 +59,7 @@ func Save(config *api.Config, cmd *cobra.Command) {
 
 // Restore loads the arguments from disk and prefills the Request
 func Restore(config *api.Config, cmd *cobra.Command) {
+	glog.V(0).Infof("Restoring")
```
Comment (bparees) [inline pkg/config/config.go]: debug?

File: pkg/config/config.go [hunk 6]
```diff
@@ -63,14 +74,31 @@ func Restore(config *api.Config, cmd *cobra.Command) {
...
+
+	envOverride := false
 	config.BuilderImage = c.BuilderImage
 	config.Source = c.Source
 	config.Tag = c.Tag
+	cmd.Flags().Visit(func(f *pflag.Flag) {
+		if f.Name == "env" {
+			envOverride = true
+		}
+	})
```
Comment (bparees) [inline pkg/config/config.go]: can't just use a cmd.Flag("env").Changed check for this?  because it's an array flag?

**Comment by bparees** [comment]
> That moment when you realise your pre-existing serialisation format is utterly inadequate 😢

yeah @coreydaley and i were discussing this.  If i had my time machine, this feature would not exist.  I'm somewhat inclined to fix it quickly here, but deprecate it going forward.  We obviously did not think through the version compatibility issues w/ storing/restoring the config.

The new struct is a good idea.

**Comment by coreydaley** [comment]
@bparees @jim-minter updated, ptal

File: pkg/config/config.go [hunk 7]
```diff
@@ -58,19 +66,40 @@ func Restore(config *api.Config, cmd *cobra.Command) {
...
+	if cmd.Flag("env").Changed {
+		envOverride = true
+	}
+
 	for name, value := range c.Flags {
-		// Do not change flags that user sets. Allow overriding of stored flags.
-		if cmd.Flag(name).Changed {
+		if envOverride {
 			continue
 		}
```
Comment (bparees) [inline pkg/config/config.go]: this is in the wrong place, isn't it?

File: pkg/config/config.go [hunk 8]
```diff
@@ -58,19 +66,40 @@ func Restore(config *api.Config, cmd *cobra.Command) {
...
 	config.Source = c.Source
 	config.Tag = c.Tag
+
+	envOverride := false
+	if cmd.Flag("env").Changed {
+		envOverride = true
+	}
+
 	for name, value := range c.Flags {
-		// Do not change flags that user sets. Allow overriding of stored flags.
```
Comment (bparees) [inline pkg/config/config.go]: jim asked to preserve this comment (and I agree)

**Comment by coreydaley** [comment]
@bparees updated, ptal

**Comment by bparees** [comment]
lgtm [merge]

**Comment by bparees** [comment]
flake https://github.com/openshift/source-to-image/issues/533
[merge]

**Pull request merged**